File: lib/java/src/main/java/io/github/brenbar/uapi/internal/types/UMockCall.java
package io.github.brenbar.uapi.internal.types;

import static io.github.brenbar.uapi.internal.validation.ValidateMockCall.validateMockCall;

import java.util.List;
import java.util.Map;

import io.github.brenbar.uapi.RandomGenerator;
import io.github.brenbar.uapi.internal.validation.ValidationFailure;

public class UMockCall implements UType {

    public static final String _MOCK_CALL_NAME = "_ext.Call_";

    public final Map<String, UType> types;

    public UMockCall(Map<String, UType> types) {
        this.types = types;
    }

    @Override
    public int getTypeParameterCount() {
        return 0;
    }

    @Override
    public List<ValidationFailure> validate(Object givenObj, Map<String, Object> select, String fn,
            List<UTypeDeclaration> typeParameters,
            List<UTypeDeclaration> generics) {
        return validateMockCall(givenObj, select, fn, typeParameters, generics, this.types);
    }

    @Override
    public Object generateRandomValue(Object blueprintValue, boolean useBlueprintValue,
            boolean includeOptionalFields, boolean randomizeOptionalFields, List<UTypeDeclaration> typeParameters,
            List<UTypeDeclaration> generics,
            RandomGenerator randomGenerator) {
        throw new UnsupportedOperationException("Not implemented");
    }

    @Override
    public String getName(List<UTypeDeclaration> generics) {
        return _MOCK_CALL_NAME;
    }

}

File: lib/ts/uapi/internal/types/UMockCall.ts
import { RandomGenerator } from 'uapi/RandomGenerator';
import { UTypeDeclaration } from 'uapi/internal/types/UTypeDeclaration';
import { ValidationFailure } from 'uapi/internal/validation/ValidationFailure';
import { UType } from 'uapi/internal/types/UType';
import { validate_mock_call } from 'uapi/internal/validation/ValidateMockCall';

export class UMockCall extends UType {
    private static _MOCK_CALL_NAME: string = '_ext.Call_';
    private types: { [key: string]: UType };

    constructor(types: { [key: string]: UType }) {
        super();
        this.types = types;
    }

    public get_type_parameter_count(): number {
        return 0;
    }

    public validate(
        given_obj: any,
        select: { [key: string]: any } | null,
        fn: string | null,
        type_parameters: UTypeDeclaration[],
        generics: UTypeDeclaration[],
    ): ValidationFailure[] {
        return validate_mock_call(given_obj, select, fn, type_parameters, generics, this.types);
    }

    public generate_random_value(
        blueprint_value: any,
        use_blueprint_value: boolean,
        include_optional_fields: boolean,
        randomize_optional_fields: boolean,
        type_parameters: UTypeDeclaration[],
        generics: UTypeDeclaration[],
        random_generator: RandomGenerator,
    ): any {
        throw new Error('Not implemented');
    }

    public get_name(generics: UTypeDeclaration[]): string {
        return UMockCall._MOCK_CALL_NAME;
    }
}
